src/workspacer.Sound/SoundManager.cs

```diff
@@ -0,0 +1,393 @@
+        }
+
+        private void ApplyToClientsWithCriteria(string deviceId, DeviceType? type, DeviceRole? role, Action<SoundClient> action)
+        {
+            IEnumerable<SoundClient> clients = null;
+            if (string.IsNullOrEmpty(deviceId))
+            {
+                clients = _clients.Where(x => x.DeviceInfo.Roles.Contains(role.Value) && x.DeviceInfo.Type == type.Value).ToList();
+            }
+            else
+            {
+                clients = _clients.Where(x => x.DeviceInfo.Id == deviceId).ToList();
+            }
+
+            foreach (var client in clients)
+            {
+                action(client);
+            }
+        }
+
+        #region System Events
+
+        public void OnDefaultDeviceChanged(EDataFlow type, ERole role, string defaultDeviceId)
+        {
+            ApplyToClientsWithCriteria(string.Empty, type.ConvertToDeviceType(), role.ConvertToDeviceRole(), (client) =>
+            {
+                _mmDeviceEnumerator.GetDevice(defaultDeviceId, out var device);
+                if (device == null)
+                {
+                    return;
+                }
+
+                client.SetDevice(device);
+                client.Refresh();
+            });
+
+#if EVENT_DEBUG
+            MessageBox.Show($"DeviceRemoved DataFlow:[{dataFlow}] DeviceRole:[{deviceRole}] DeviceId:[{defaultDeviceId}]");
+#endif
+        }
+
+        public void OnDeviceStateChanged(string deviceId, uint newState)
+        {
+            ApplyToClientsWithCriteria(deviceId, null, null, (client) => client.Refresh() );
+
+            DeviceChanged?.Invoke(new DeviceInfo { Id = deviceId }, newState);
+#if EVENT_DEBUG
+            MessageBox.Show($"DeviceStateChanged deviceId:[{deviceId}] newState:[{newState}]");
+#endif
+        }
+
+        public void OnDeviceAdded(string deviceId)
+        {
+            ApplyToClientsWithCriteria(deviceId, null, null, (client) => client.Refresh());
+
+            DeviceAdded?.Invoke(new DeviceInfo { Id = deviceId });
+#if EVENT_DEBUG
+            MessageBox.Show($"DeviceAdded DeviceId:[{deviceId}]");
+#endif
+        }
+
+        public void OnDeviceRemoved(string deviceId)
+        {
+            ApplyToClientsWithCriteria(deviceId, null, null, (client) => client.Refresh());
+
+            DeviceRemoved?.Invoke(new DeviceInfo { Id = deviceId });
+#if EVENT_DEBUG
+            MessageBox.Show($"DeviceRemoved DeviceId:[{deviceId}]");
+#endif
+        }
+
+        public void OnPropertyValueChanged(string deviceId, PROPERTYKEY propertyKey)
+        {
+            ApplyToClientsWithCriteria(deviceId, null, null, (client) => client.Refresh());
+
+            DevicePropertyValueChanged?.Invoke(new DeviceInfo { Id = deviceId }, propertyKey);
+#if EVENT_DEBUG
+            MessageBox.Show($"DeviceStateChanged deviceId:[{deviceId}] propertyKey:[{propertyKey}]");
+#endif
+        }
+        #endregion
+    }
+
+    public class SoundClient : IAudioEndpointVolumeCallback
```

Comment:
I see the `SoundClient` class being exposed to the "outside world" through the `SoundManager` class.

As such, I think it would be nice to put in its own file. As a general rule, I think that's always nice.